### PR TITLE
HTML-748: Obs element: support attribute "answerDrugIds"

### PR DESCRIPTION
**Pull request by mogoodrich**
## Issue I worked on

see https://issues.openmrs.org/browse/HTML-

## Checklist: I completed these to help reviewers :)

- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java [hunk 1]
```diff
@@ -797,11 +801,37 @@ private void prepareWidgets(FormEntryContext context, Map<String, String> parame
 							widget.setDisplayTemplate("{{name}}");
 						}
 						if (existingObs != null && existingObs.getValueDrug() != null) {
+							// TODO: not quite sure how this works?
 							widget.setInitialValue(new Option(existingObs.getValueDrug().getName(),
 							        existingObs.getValueDrug().getDrugId().toString(), true));
 						}
 						valueWidget = widget;
-						
+					}
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:809]: I would think that this should be within the above else block.  eg. if parameters.get("answerDrugs") != null, then this is a drug widget.  By default, make this an autocomplete, but if answerDrugIds is passed in, then make it a dropdown.

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java [hunk 2]
```diff
@@ -797,11 +801,37 @@ private void prepareWidgets(FormEntryContext context, Map<String, String> parame
...
+							String s = st.nextToken().trim();
+							Drug drug = HtmlFormEntryUtil.getDrug(s);
+							if (drug == null) {
+								throw new IllegalArgumentException("Cannot find drug for value " + s
+								        + " in answerDrugs attribute value. Parameters: " + parameters);
+							}
+							answerDrugs.add(drug);
+						}
+
+						valueWidget = buildDropdownWidget(size);
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:823]: If this doesn't automatically create an "empty" option, wed need to make sure one exists.

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java [hunk 3]
```diff
@@ -797,11 +801,37 @@ private void prepareWidgets(FormEntryContext context, Map<String, String> parame
...
+						for (int i = 0; i < answerDrugs.size(); i++) {
+							String label = answerLabels != null && answerLabels.size() > i ? answerLabels.get(i)
+							        : answerDrugs.get(i).getName();
+							((SingleOptionWidget) valueWidget).addOption(
+							    new Option(label, "Drug:" + answerDrugs.get(i).getDrugId().toString(), false));
+						}
+
+						if (existingObs != null && existingObs.getValueDrug() != null) {
+							valueWidget.setInitialValue("Drug:" + existingObs.getValueDrug().getDrugId().toString());
+						}
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:834]: This is all super ugly, but I assume this is mostly a copy and paste job from elsewhere in this class?

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java [hunk 4]
```diff
@@ -797,11 +801,37 @@ private void prepareWidgets(FormEntryContext context, Map<String, String> parame
...
+							answerDrugs.add(drug);
+						}
+
+						valueWidget = buildDropdownWidget(size);
+
+						for (int i = 0; i < answerDrugs.size(); i++) {
+							String label = answerLabels != null && answerLabels.size() > i ? answerLabels.get(i)
+							        : answerDrugs.get(i).getName();
+							((SingleOptionWidget) valueWidget).addOption(
+							    new Option(label, "Drug:" + answerDrugs.get(i).getDrugId().toString(), false));
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:829]: What is this all about?  The value in the dropdown is "Drug:" followed by the drug id?  I guess if this is how things are already designed to work, so be it.

File: api/src/test/resources/org/openmrs/module/htmlformentry/include/obsDrugFormWithDropdown.xml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<htmlform>
+    Date: <encounterDate/>
+    Location: <encounterLocation/>
+    Provider: <encounterProvider role="Provider"/>
+
+    Allergic to drug: <obs conceptId="1000" answerDrugIds="3cfcf118-931c-46f7-8ff6-7b876f0d4202,05ec820a-d297-44e3-be6e-698531d9dd3f"/>
```
Comment (mseaton) [inline api/src/test/resources/org/openmrs/module/htmlformentry/include/obsDrugFormWithDropdown.xml:6]: This is going to get ugly when put into practice on forms, if the lists of drugs are long...

I suppose introducing a nested tag is not easy.

**Review by mseaton — APPROVED** [review]
Ugly, but no uglier than the rest of ObsSubmissionElement I guess.  Lots of white-space changes I had to suppress in the PR.  A few comments you can address if appropriate.

File: api/src/test/resources/org/openmrs/module/htmlformentry/include/obsDrugFormWithDropdown.xml (hunk 5 above)
Comment (mogoodrich) [reply api/src/test/resources/org/openmrs/module/htmlformentry/include/obsDrugFormWithDropdown.xml:6]: Yeah, what we really need is mapping functionality like concept mapping... you mean a nested tag where there's a tag per drug option?  Yes, not sure that would be easy/doable... and this follows the existing pattern for answerConceptId (except, of course, mappings are supported in that case).

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java (hunk 2 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:823]: Yep, that's part of what is does.

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java (hunk 1 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:809]: I could see that, but I'd prefer not to require someone to put both answerDrugs and answerDrugIds, seems redundant.

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java (hunk 4 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:829]: Yes, tthat's the pattern.

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java (hunk 3 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:834]: Yeah... There's actually a more generic place where I was considering adding this functionality (ie something that mapped a Drug to "Drug:123" but unfortunately at this point with ObsSubmissionElement I think the more we constraint our changes to the code to certain areas the better.

**Review by cioan — APPROVED** [review]
LGTM, thanks @mogoodrich !

**Comment by mogoodrich** [comment]
Thanks @cioan  and @mseaton !

**Pull request merged**